Firestore/core/test/firebase/firestore/local/counting_query_engine.h

```diff
@@ -0,0 +1,204 @@
+  bool index_free_;
+
+  std::unique_ptr<LocalDocumentsView> local_documents_;
+  std::unique_ptr<WrappedMutationQueue> mutation_queue_;
+  std::unique_ptr<WrappedRemoteDocumentCache> remote_documents_;
+
+  int mutations_read_by_query_ = 0;
+  int mutations_read_by_key_ = 0;
+  int documents_read_by_query_ = 0;
+  int documents_read_by_key_ = 0;
+};
+
+/** A MutationQueue that counts document reads. */
+class WrappedMutationQueue : public MutationQueue {
+ public:
+  WrappedMutationQueue(MutationQueue* subject,
+                       CountingQueryEngine* query_engine)
+      : subject_(subject), query_engine_(query_engine) {
+  }
+
+  void Start() override;
+
+  bool IsEmpty() override;
+
+  void AcknowledgeBatch(const model::MutationBatch& batch,
+                        const nanopb::ByteString& stream_token) override;
+
+  model::MutationBatch AddMutationBatch(
+      const Timestamp& local_write_time,
+      std::vector<model::Mutation>&& base_mutations,
+      std::vector<model::Mutation>&& mutations) override;
+
+  void RemoveMutationBatch(const model::MutationBatch& batch) override;
+
+  std::vector<model::MutationBatch> AllMutationBatches() override;
+
+  std::vector<model::MutationBatch> AllMutationBatchesAffectingDocumentKeys(
+      const model::DocumentKeySet& document_keys) override;
+
+  std::vector<model::MutationBatch> AllMutationBatchesAffectingDocumentKey(
+      const model::DocumentKey& key) override;
+
+  std::vector<model::MutationBatch> AllMutationBatchesAffectingQuery(
+      const core::Query& query) override;
+
+  absl::optional<model::MutationBatch> LookupMutationBatch(
+      model::BatchId batch_id) override;
+
+  absl::optional<model::MutationBatch> NextMutationBatchAfterBatchId(
+      model::BatchId batch_id) override;
+
+  model::BatchId GetHighestUnacknowledgedBatchId() override;
+
+  void PerformConsistencyCheck() override;
+
+  nanopb::ByteString GetLastStreamToken() override;
+
+  void SetLastStreamToken(nanopb::ByteString stream_token) override;
+
+ private:
+  MutationQueue* subject_;
+  CountingQueryEngine* query_engine_;
+};
+
+/** A RemoteDocumentCache that counts document reads. */
+class WrappedRemoteDocumentCache : public RemoteDocumentCache {
+ public:
+  WrappedRemoteDocumentCache(RemoteDocumentCache* subject,
+                             CountingQueryEngine* query_engine)
+      : subject_(subject), query_engine_(query_engine) {
+  }
+
+  void Add(const model::MaybeDocument& document,
+           const model::SnapshotVersion& read_time) override;
+
+  void Remove(const model::DocumentKey& key) override;
+
+  absl::optional<model::MaybeDocument> Get(
+      const model::DocumentKey& key) override;
+
+  model::OptionalMaybeDocumentMap GetAll(
+      const model::DocumentKeySet& keys) override;
+
+  model::DocumentMap GetMatching(
+      const core::Query& query,
+      const model::SnapshotVersion& since_read_time) override;
+
+ private:
+  RemoteDocumentCache* subject_;
```

Comment:
All primitive type members (including raw pointers) should have initializers. Use `= nullptr` for both of these.

Reply:
Done